File: javascript/src/RegularExpression.ts
import Argument from './Argument'
import ParameterType from './ParameterType'
import type ParameterTypeRegistry from './ParameterTypeRegistry'
import TreeRegexp from './TreeRegexp'
import type { Expression } from './types'

export default class RegularExpression implements Expression {
  private readonly treeRegexp: TreeRegexp

  constructor(
    public readonly regexp: RegExp,
    private readonly parameterTypeRegistry: ParameterTypeRegistry
  ) {
    this.treeRegexp = new TreeRegexp(regexp)
  }

  public match(text: string): readonly Argument[] | null {
    const group = this.treeRegexp.match(text)
    if (!group) {
      return null
    }

    const parameterTypes = this.treeRegexp.groupBuilder.children.map((groupBuilder) => {
      const parameterTypeRegexp = groupBuilder.source

      const parameterType = this.parameterTypeRegistry.lookupByRegexp(
        parameterTypeRegexp,
        this.regexp,
        text
      )
      return (
        parameterType ||
        new ParameterType(
          undefined,
          parameterTypeRegexp,
          String,
          (s) => (s === undefined ? null : s),
          false,
          false
        )
      )
    })

    return Argument.build(group, parameterTypes)
  }

  get source(): string {
    return this.regexp.source
  }
}

File: dart/lib/src/regular_expression.dart
import 'package:cucumber_expressions/src/argument.dart';
import 'package:cucumber_expressions/src/expression.dart';
import 'package:cucumber_expressions/src/parameter_type.dart';
import 'package:cucumber_expressions/src/parameter_type_lookup.dart';
import 'package:cucumber_expressions/src/parameter_type_registry.dart';
import 'package:cucumber_expressions/src/tree_regexp.dart';

class RegularExpression implements Expression {
  RegularExpression(this.regexp, this._parameterTypeRegistry)
      : _treeRegexp = TreeRegexp(regexp);

  final RegExp regexp;
  final ParameterTypeRegistry _parameterTypeRegistry;
  final TreeRegexp _treeRegexp;

  @override
  List<Argument<Object?>>? match(String text) {
    final group = _treeRegexp.match(text);
    if (group == null) {
      return null;
    }

    final parameterTypes =
        _treeRegexp.groupBuilder.children.map((groupBuilder) {
      final parameterTypeRegexp = groupBuilder.source;

      final parameterType = lookupParameterTypeByRegexp(
        _parameterTypeRegistry,
        parameterTypeRegexp,
        regexp,
        text,
      );
      return parameterType ??
          ParameterType<String?>(
            null,
            parameterTypeRegexp,
            'String',
            (s) => s.first,
            useForSnippets: false,
            preferForRegexpMatch: false,
          );
    }).toList();

    return buildArguments(group, parameterTypes);
  }

  @override
  String get source => regexp.pattern;
}
